File: app/resources/tcltk/hellotcltkdemo.py
#!/usr/bin/env python -t

from __future__ import (absolute_import, division, print_function,
  unicode_literals)

import sys, os, logging, inspect
from future.builtins import (ascii, filter, hex, map, oct, zip, str)

try:
  import tkinter as tk
except ImportError as exc:
  print(repr(exc))
  import Tkinter as tk

def userifc_version():
  import platform

  return "(Python {0}) Tk {1}".format(platform.python_version(),
    tk.TkVersion)

class HelloTcltkDemo(object):
  '''Description:: '''

  def __init__(self, app=None):
    '''Construct object'''

    super(HelloTcltkDemo, self).__init__()
    
    self.app, self.widgets = app, {}
    
    self.widgets['frame1'] = tk.Frame(self.app)
    self.widgets['label1'] = tk.Label(self.widgets['frame1'], text='label1')
    self.widgets['button1'] = tk.Button(self.widgets['frame1'], 
      text='button1')
    self.widgets['textview1'] = tk.Text(self.widgets['frame1'], height=3, 
      width=30)
    self.widgets['frame1'].pack()
    
    for widget in map(self.widgets.get, ['label1', 'button1', 'textview1']):
      widget.pack()
    
    self.widgets['dialog1'] = tk.Toplevel()
    self.widgets['dialog1'].title('dialog1')
    self.widgets['frameDialog1'] = tk.Frame(self.widgets['dialog1'])
    self.widgets['txt1'] = tk.StringVar()
    self.widgets['entry1'] = tk.Entry(master=self.widgets['frameDialog1'],
      width=25, textvariable=self.widgets['txt1'])
    self.widgets['entry1'].pack()
    self.widgets['frameDialog1'].pack()
    self.widgets['dialog1'].withdraw()
    
    self.widgets['dialog1'].bind('<Destroy>', self.dialog1_destroyed_cb)
    self.widgets['button1'].bind('<Button>', self.button1_clicked_cb)
    self.widgets['txt1'].trace('w', self.entry1_textChanged_cb)
    self.widgets['entry1'].bind('<Return>', self.entry1_returnPressed_cb)
    
    self.app.title('frame1')
    self.widgets['frame1'].lift()

  def dialog1_destroyed_cb(self, dialog1, callback_data=None):
    '''Callback for dialog1 destroyed'''

    self.app.quit()

  def button1_clicked_cb(self, button1, callback_data=None):
    '''Callback for button1 clicked'''
    
    #try:
    #  import tkSimpleDialog as simpledialog
    #except ImportError as exc:
    #  print(repr(exc))
    #  import tkinter.simpledialog as simpledialog
    #data = simpledialog.askstring('dialog1', '', 
    #  parent=self.widgets['frame1'])

    self.widgets['dialog1'].deiconify()
    self.app.withdraw()
    self.widgets['entry1'].focus_set()
    self.widgets['entry1'].delete(0, tk.END)

  def entry1_textChanged_cb(self, entry1, extra, callback_data=None):
    '''Callback for entry1 textChanged'''

    self.widgets['textview1'].delete(1.0, tk.END)

  def entry1_returnPressed_cb(self, entry1, callback_data=None):
    '''Callback for entry1 returnPressed'''

    self.app.deiconify()
    self.widgets['dialog1'].withdraw()
    self.widgets['textview1'].delete(1.0, tk.END)
    self.widgets['textview1'].insert(tk.INSERT,
      'Hello, {}.'.format(self.widgets['entry1'].get()))

def main(argv=None):
  app = tk.Tk()
  
  pretext = '{0} GUI\n'.format(userifc_version())
  gui = HelloTcltkDemo(app)
  gui.widgets['textview1'].delete(1.0, tk.END)
  gui.widgets['textview1'].insert(tk.INSERT, pretext)
  
  app.mainloop()
  return 0

if '__main__' == __name__:
  sys.exit(main(sys.argv[1:]))


# (one-time): pip install --user tkinter
# python hellotcltkdemo.py

File: app/resources/tcltk/hellotcltkdemo.rb
#!/usr/bin/env ruby -w

=begin rdoc
Description:: Brief description
Version:: $Id:$
=end
require 'English'
require 'rubygems'
require 'tk'

class HelloTcltkDemo
  attr_reader :widgets, :app
  
  def initialize(app=nil)
    @app, @widgets = app, {}
    
    @widgets['frame1'] = TkFrame.new(@app)
    @widgets['label1'] = TkLabel.new(@widgets['frame1'], text: 'label1')
    @widgets['button1'] = TkButton.new(@widgets['frame1'], text: 'button1')
    @widgets['textview1'] = TkText.new(@widgets['frame1'], height: 3, 
      width: 30)
    @widgets['frame1'].pack
    
    %w(label1 button1 textview1).each { |name|
      @widgets[name].pack }
    
    @widgets['dialog1'] = TkToplevel.new
    @widgets['dialog1'].title 'dialog1'
    @widgets['frameDialog1'] = TkFrame.new(@widgets['dialog1'])
    @widgets['txt1'] = TkVariable.new
    @widgets['entry1'] = TkEntry.new @widgets['frameDialog1'], width: 25,
      textvariable: @widgets['txt1']
    @widgets['entry1'].pack
    @widgets['frameDialog1'].pack
    @widgets['dialog1'].withdraw
    
    @widgets['dialog1'].bind('Destroy', proc{ |*args|
      dialog1_destroy_cb(*args) })
    @widgets['button1'].bind('Button', proc{ |*args| 
      button1_clicked_cb(*args) })
    @widgets['txt1'].trace('w', proc{ |*args|
      entry1_textChanged_cb(*args) })
    @widgets['entry1'].bind('Return', proc{ |*args|
      entry1_returnPressed_cb(*args) })
    
    @app.title 'frame1'
  end
  
  private
  def dialog1_destroy_cb(widget, callback_data=nil)
    @app.destroy
  end
  
  def button1_clicked_cb(widget, callback_data=nil)
    @widgets['dialog1'].deiconify
    @app.withdraw
    @widgets['entry1'].set_focus
    #@widgets['entry1'].delete(0, 'end')
    @widgets['entry1'].value = ''
  end
  
  def entry1_textChanged_cb(widget, extra, callback_data=nil)
    #@widgets['textview1'].delete(0)
    @widgets['textview1'].value = ''
  end
  
  def entry1_returnPressed_cb(widget, callback_data=nil)
    @app.deiconify
    @widgets['dialog1'].withdraw
    #@widgets['textview1'].delete(0)
    @widgets['textview1'].value = ''
    @widgets['textview1'].insert('end',
      'Hello, %s.' % @widgets['entry1'].get)
  end
end # end class HelloTcltkDemo

if $PROGRAM_NAME == __FILE__
  app = TkRoot.new
  
  pretext = '(Ruby %s) Tk %s GUI' % [RUBY_VERSION, Tk::TK_VERSION]
  gui = HelloTcltkDemo.new app
  #gui.widgets['textview1'].delete(0)
  gui.widgets['textview1'].value = ''
  gui.widgets['textview1'].insert('end', pretext)
  
  Tk.mainloop

  exit 0
end


# (one-time): gem install --user-install tk
# ruby hellotcltkdemo.rb
